app_dart/lib/src/request_handlers/refresh_cirrus_status.dart

```diff
@@ -37,60 +42,91 @@ class RefreshCirrusStatus extends ApiRequestHandler<Body> {
+    @visibleForTesting this.gitHubBackoffCalculator = twoSecondLinearBackoff,
   })  : datastoreProvider =
             datastoreProvider ?? DatastoreService.defaultProvider,
+        assert(branchHttpClientProvider != null),
+        assert(gitHubBackoffCalculator != null),
         super(config: config, authenticationProvider: authenticationProvider);
 
   final DatastoreServiceProvider datastoreProvider;
+  final HttpClientProvider branchHttpClientProvider;
+  final GitHubBackoffCalculator gitHubBackoffCalculator;
 
   @override
   Future<Body> get() async {
     final DatastoreService datastore = datastoreProvider();
     final GraphQLClient client = await config.createCirrusGraphQLClient();
+    final GitHub github = await config.createGitHubClient();
+    const RepositorySlug slug = RepositorySlug('flutter', 'flutter');
+    final Stream<Branch> branches = github.repositories.listBranches(slug);
+    final List<String> regExps = await loadBranchRegExps(
+        branchHttpClientProvider, log, gitHubBackoffCalculator);
 
-    await for (FullTask task
-        in datastore.queryRecentTasks(taskName: 'cirrus', commitLimit: 15)) {
-      final String sha = task.commit.sha;
-      final String existingTaskStatus = task.task.status;
-      log.debug(
-          'Found Cirrus task for commit $sha with existing status $existingTaskStatus');
-      final List<String> statuses = <String>[];
-      const String name = 'flutter';
-
-      for (Map<String, dynamic> runStatus
-          in await queryCirrusGraphQL(sha, client, log, name)) {
-        final String status = runStatus['status'] as String;
-        final String taskName = runStatus['name'] as String;
-        log.debug('Found Cirrus build status for $sha: $taskName ($status)');
-        statuses.add(status);
+    await for (Branch branch in branches) {
+      if (!regExps
+          .any((String regExp) => RegExp(regExp).hasMatch(branch.name))) {
+        continue;
       }
+      await for (FullTask task in datastore.queryRecentTasks(
+          taskName: 'cirrus', commitLimit: 15, branch: branch.name)) {
+        final String sha = task.commit.sha;
+        final String existingTaskStatus = task.task.status;
 
-      String newTaskStatus;
-      if (statuses.isEmpty) {
-        newTaskStatus = Task.statusNew;
-      } else if (statuses.any(kCirrusFailedStates.contains)) {
-        newTaskStatus = Task.statusFailed;
-        task.task.endTimestamp = DateTime.now().millisecondsSinceEpoch;
-      } else if (statuses.any(kCirrusInProgressStates.contains)) {
-        newTaskStatus = Task.statusInProgress;
-      } else {
-        newTaskStatus = Task.statusSucceeded;
-        task.task.endTimestamp = DateTime.now().millisecondsSinceEpoch;
-      }
+        log.debug(
+            'Found Cirrus task for branch ${branch.name}, commit $sha, with existing status $existingTaskStatus');
+
+        final List<String> statuses = <String>[];
```

Comment:
Seems like we have too much logic in a single method, can we split the functionality into multiple methods? This will also simplify tests for specific functionality.

Reply:
Added two sub functions: _getStatuses and _getNewTaskStatus to simply.